### PR TITLE
abort write transactions

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

This is an internal PR for hot backup, so that it can optionally abort ongoing write transactions and ensure a backup can be taken.
Also improves killability of AQL queries.

- [x] Bug-Fix for *devel-branch* (i.e. no need for backports?)
- [x] The behavior change can be verified via automatic tests

### Testing & Verification

This PR adds tests that were used to verify all changes:

- [x] Added new **integration tests** (i.e. in shell_server / shell_server_aql)

https://172.16.10.101/view/PR/job/arangodb-matrix-pr/6763/

### CODE REVIEW

**Comment by jsteemann** [comment]
https://172.16.10.101/view/PR/job/arangodb-matrix-pr/6773/

File: arangod/Aql/DistributeExecutor.cpp [hunk 1]
```diff
@@ -224,6 +232,9 @@ std::pair<ExecutionState, bool> ExecutionBlockImpl<DistributeExecutor>::getBlock
 
   while (buf.size() < atMost) {
     if (_index == _buffer.size()) {
+      if (getQuery().killed()) {
+        THROW_ARANGO_EXCEPTION(TRI_ERROR_QUERY_KILLED);
+      }
```
Comment (graetzer) [inline arangod/Aql/DistributeExecutor.cpp]: seems expensive to check this inside the loop? shouldn't this be checkd on the DBServers as well

File: arangod/Aql/QueryList.cpp [hunk 2]
```diff
@@ -186,32 +187,36 @@ void QueryList::remove(Query* query) {
...
       << "killing AQL query " << id << " '" << query->queryString() << "'";
 
   query->kill();
-  return TRI_ERROR_NO_ERROR;
+  return Result();
 }
 
 /// @brief kills all currently running queries
-uint64_t QueryList::killAll(bool silent) {
+uint64_t QueryList::kill(std::function<bool(Query*)> const& filter, bool silent) {
```
Comment (graetzer) [inline arangod/Aql/QueryList.cpp]:
```suggestion
uint64_t QueryList::kill(std::function<bool(Query&)> const& filter, bool silent) {
```

File: arangod/Aql/DistributeExecutor.cpp (hunk 1 above)
Comment (jsteemann) [reply arangod/Aql/DistributeExecutor.cpp]: The cost should be negligible here, as the check is only performed when a new block is fetched anyway, but not for every "row".

File: arangod/Aql/QueryList.cpp (hunk 2 above)
Comment (jsteemann) [reply arangod/Aql/QueryList.cpp]: Changed accordingly

**Comment by jsteemann** [comment]
https://172.16.10.101/view/PR/job/arangodb-matrix-pr/6822/

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/6833/

**Pull request merged**